miden-tx/src/executor/mod.rs

```diff
@@ -0,0 +1,104 @@
+use super::{
+    AccountId, DataStore, NoteOrigin, PreparedTransaction, ProgramAst, RecAdviceProvider,
+    TransactionComplier, TransactionExecutorError, TransactionWitness,
+};
+
+/// The [TransactionExecutor] has the following responsibilities:
+/// - Fetch the data required to execute a transaction from the [DataStore].
+/// - Compile the transaction into a [Program] using the [TransactionComplier].
+/// - Execute the transaction [Program] and create a [TxWitness] using the [AdviceRecorder].
+///
+/// The [TransactionExecutor] is generic over the [DataStore] which allows it to be used
+/// with different data backend implementations.
+///
+/// The [TransactionExecutor] `execute_transaction` method is the main entry point for the executor
+/// and produces a [TxWitness] for the transaction. The [TxWitness] is then used to by the prover
+/// to generate a proof for the transaction.
+pub struct TransactionExecutor<D: DataStore> {
+    compiler: TransactionComplier,
+    data_store: D,
+}
+
+impl<D: DataStore> TransactionExecutor<D> {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Creates a new [TransactionExecutor] instance with the specified [DataStore].
+    pub fn new(data_store: D) -> Self {
+        let compiler = TransactionComplier::new();
+        Self {
+            compiler,
+            data_store,
+        }
+    }
+
+    // MODIFIERS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a mutable reference to the [TransactionComplier] owned by this executor.
+    pub fn compiler(&mut self) -> &mut TransactionComplier {
+        &mut self.compiler
+    }
```

Comment:
One other thing I'm wondering about is whether we should expose transaction compiler like this, or just have a set of pass-through methods on the executor. For example, public interface of transaction executor could look something like this:

```Rust
impl<D: DataStore> TransactionExecutor<D> {

    pub fn load_account(
        &mut self,
        account_id: AccountId,
    ) -> Result<AccountCode, TransactionError> {
        let account_code = self.store.get_account_code(account_id)?;
        self.compiler.load_account(account_id, account_code)
    }

    pub fn load_account_interface(
        &mut self,
        account_id: AccountId,
        procedures: Vec<Digest>,
    ) -> Option<Vec<Digest>> {
        self.compiler.load_account_interface(account_id, procedures)
    }

    pub fn compile_note_script(
        &mut self,
        note_script_ast: ProgramAst,
        target_account_procs: Vec<NoteTarget>,
    ) -> Result<NoteScript, TransactionError> {
        self.compiler.compile_note_script(note_script_ast, target_account_procs)
    }

    pub fn execute_transaction(
        &mut self,
        account_id: AccountId,
        block_ref: u32,
        note_hashes: &[Digest],
        tx_script: Option<ProgramAst>,
    ) -> Result<TransactionWitness, TransactionExecutorError> {
        ...
    }
}
```

And I'm actually not sure if we need to expose `load_account_interface()` and `compile_note_script()` at all as it is not clear to me how these will be used yet (it might be that we'll use them as a part of bigger function - e.g., `create_note()`).

Reply:
I've added them for now.  I think we can review if they're needed once the transaction pipeline is more fleshed out.